airbyte-integrations/connectors/source-kafka/src/main/java/io/airbyte/integrations/source/kafka/generator/GeneratorImpl.java

```diff
@@ -0,0 +1,103 @@
+ */
+
+package io.airbyte.integrations.source.kafka.generator;
+
+import com.google.common.collect.AbstractIterator;
+import io.airbyte.commons.util.AutoCloseableIterator;
+import io.airbyte.commons.util.AutoCloseableIterators;
+import io.airbyte.integrations.source.kafka.converter.Converter;
+import io.airbyte.integrations.source.kafka.mediator.KafkaMediator;
+import io.airbyte.integrations.source.kafka.state.StateHelper;
+import io.airbyte.protocol.models.v0.AirbyteMessage;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.TopicPartition;
+
+public class GeneratorImpl<V> implements Generator {
+
+  private final KafkaMediator<V> mediator;
+  private final Converter<V> converter;
+  private final int maxRecords;
+  private final int maxRetries;
+
+  public GeneratorImpl(KafkaMediator<V> mediator, Converter<V> converter, int maxRecords, int maxRetries) {
+    this.mediator = mediator;
+    this.converter = converter;
+    this.maxRecords = maxRecords;
+    this.maxRetries = maxRetries;
+  }
+
+  @Override
+  final public AutoCloseableIterator<AirbyteMessage> read() {
+
+    return AutoCloseableIterators.fromIterator(new AbstractIterator<>() {
+
+      private int totalRead = 0;
+      private final Queue<AirbyteMessage> pendingMessages = new LinkedList<>();
+
+      @Override
+      protected AirbyteMessage computeNext() {
+        if (this.pendingMessages.isEmpty()) {
+          if (this.totalRead < GeneratorImpl.this.maxRecords) {
+            List<ConsumerRecord<String, V>> batch = pullBatchFromKafka();
+            if (!batch.isEmpty()) {
+              this.totalRead += batch.size();
+              this.pendingMessages.addAll(convertToAirbyteMessagesWithState(batch));
+            }
+          } else {
+            return endOfData();
+          }
+        }
+
+        // If no more pending kafka records, close iterator
+        if (this.pendingMessages.isEmpty()) {
+          return endOfData();
+        } else {
+          return pendingMessages.poll();
+        }
+      }
+
+      private List<AirbyteMessage> convertToAirbyteMessagesWithState(List<ConsumerRecord<String, V>> batch) {
+        final Set<TopicPartition> partitions = new HashSet<>();
+        final List<AirbyteMessage> messages = new ArrayList<>();
+
+        for (ConsumerRecord<String, V> entry : batch) {
+          final var topic = entry.topic();
+          final var partition = entry.partition();
+          final var value = entry.value();
+          partitions.add(new TopicPartition(topic, partition));
+          messages.add(
+              new AirbyteMessage()
+                  .withType(AirbyteMessage.Type.RECORD)
+                  .withRecord(GeneratorImpl.this.converter.convertToAirbyteRecord(topic, value))

```

Comment:
The reason is that this is a inner anonymous class. In situations like that I always namespace outer variables properly to avoid weird issues with redefinitions. `GeneratorImpl.this.converter` means the converter property that belongs to the current instance of the outer class. I know it's verbose, but I prefer this "defensive" way of refesring to outer variables.

Reply:
OK, understood, tnx 👍🏻